docs/proposals/multiple-sources-for-applications-ui.md

```diff
@@ -0,0 +1,225 @@
+
+2. History and Rollback is disabled for multi-source applications. The button is disabled. Jorge has submitted a PR for 
+rollback which includes controller and UI changes [here](https://github.com/argoproj/argo-cd/pull/14124).
+
+   <img height="40%" width="40%" src="images/history-and-rollback-button.png"/>
+
+3. The New Application dialog currently only allows users to provide one source.
+
+Thus, multiple source applications are not considered first class citizens in the UI.
+
+Note, see the Open Questions section for concerns regarding the priority or value of some of the above changes.
+
+## Motivation
+
+The motivation behind this change is to add a more complete story for the multiple source feature. The UI should support 
+the creation of multiple source applications, and also support the viewing and editing of parameters from all sources. The three
+points in the summary above are the base or core changes that need to be addressed.
+
+### Goals
+
+The goals of the proposal are:
+
+- Provide first-class support of multiple sources for applications in the UI (eg. address the aforementioned restrictions)
+- Outline stages of implementation that will help ease PR review, and reduce the risk of introducing regressions/issues.
+
+
+### Non-goals
+* The design changes for the Argo CD CLI is beyond the scope of this proposal (The server APIs can probably be reused)
+
+## Proposal
+
+As mentioned in the previous summary section, the application source parameters are surfaced in the UI in three locations.
+The Resource details pages, specifically, the Summary and Parameters tabs, the deployment history, and the Application
+Create panel page. These pages should be updated.
+
+### Resource Details
+
+The following describes the current behavior and proposed changes for the Summary tab and the Parameters Tab.
+
+#### i) Summary Tab
+
+_Current Behavior:_ 
+
+The current Summary tab includes source-related information, including the repository. For example, in Figure 1 below,
+the REPO URL and PATH.
+
+<img height="50%" width="50%" src="images/current-summary-tab.png"/>
+
+Figure 1: The current Summary tab
+
+_Proposed Change:_
+
+To support multiple sources, the source-related information, from a single-source-based design, will be ‘pulled out’ 
+and put into a new tab called **Sources**, and it will be combined with the **Parameters** tab (more details following). 
+The new **Sources** tab will allow users to view all the information related to each source, including the repo URL 
+and path, chart and revision for Helm, etc. 
+
+The view should show one source at a time (similar to what the UI is doing now, which only shows one source), but with 
+widgets to allow users to cycle (via pagination or combo selector?) through each source. There are API calls to retrieve 
+the data for each source.
+
+<img height="50%" width="50%" src="images/new-sources-tab.png"/>
```

Comment:
I'm not a huge fan of this specific rendering.

- Generally this specific widget is more commonly used at the bottom of a view (sometimes both top and bottom).
- I don't have a better proposal at this time.
- Sometimes drop-downs are used for this purpose, although there are definite downsides. 

How many are likely? Just 3, or are 10, and 100 conceivable/likely?

Reply:
Hi @jsoref, I think there will be advantages and disadvantages.  In the appendix, I talked about collapsible cards where each card houses the source/input parameters. But it can get unwieldy if we try to support N-number of sources.

Maybe it's could end up being a combination of collapsible cards and pagination.  Note that there is some time required to load the source details.  So potentially, we can defer loading until the page is turned.  If all cards are in the one tab, then all sources need to be loaded.

Reply:
I think I'm slightly more inclined to a vertical list of things on the left that one can click on, akin to: github.com/settings:
<img width="286" alt="image" src="https://github.com/argoproj/argo-cd/assets/2119212/67efa9a8-aa22-4e76-b2e8-bf925e235a57">

It's also more or less how IntelliJ and some other things manage...

But, I'm not wed to anything... This is more of a seeking scenarios.

Reply:
I think it would be difficult to add enough information about the sources in the vertical view without making each item a bigger block with multiple fields. It might be good to have 10 sources per page and a collapsible card / pop-up modal for each source.

Reply:
I'd go for consistency with what we have now.  It parallels the List view like I mentioned above, and it's similar to the layout in the Settings view.  However, these cards are expandable.

Four sources, with the cards collapsed.

<img width="975" alt="FourSources-Collapsed" src="https://github.com/argoproj/argo-cd/assets/7726022/46631eab-da75-4fcf-9059-d65da5d1f0d8">

The third source is expanded.

<img width="983" alt="ThirdExpanded" src="https://github.com/argoproj/argo-cd/assets/7726022/4ce9ae9e-fbc5-43fb-ab03-d53e133aadde">

The pagination is needed in case there are those users who have many, many sources.

Reply:
I'm really 👎  on the pagination. As long as only one source is expanded, how long / bad is the scrolling behavior?

Also, I'd rather make the URL more prominent and the kind _less_.

Reply:
The issue is with Helm and possibly plugin-type sources. If there are many parameters, one source card can be very long (tall).  So, in the figure above, if the first Helm source is expanded, you will have to scroll a long way down to see the other source cards. It'll be even longer if all cards are expanded.  So it can quickly get unwieldy and unusable. So this usability issue can happen even for 4 sources.

See [docs/proposals/images/helm-parameter-list.png](https://github.com/argoproj/argo-cd/pull/17108/files#diff-53ce2f5f3c4e384ad977725d2c9b02fef192d8e7201b0ab926bbf461948d4bcb)

When users visit the page, all source cards should probably be initially collapsed.

Reply:
The title of the card can be the Repo URL.  We can fine-tune this as part of the implementation PR?

<img width="973" alt="RepoUrlAsTitle" src="https://github.com/argoproj/argo-cd/assets/7726022/3c248dd1-bca4-4e12-a8bc-2f45a65c9599">

Reply:
A few things:
1. the section (repo url / path / plugin source) should remain on screen while you're scrolling through its content
2. headings within it (`Directory`, `Parameters`) should be collapsible
3. ideally they should remain visible until their section scrolls off. See the GitHub Pull Request Reviews add-on for Visual Studio Code for an amazing implementation:
    https://github.com/argoproj/argo-cd/assets/2119212/df478b54-d302-4d5f-868d-ca1478d2b8d2

Reply:
Are source numbers useful/necessary? Can someone do things with them? I suppose as anchors if I wanted to send a link to someone to look at the configuration for source n, that might be useful.

If there's no other particular value in a source number, then I probably wouldn't bother.

I'd probably be more interested in a filters/search thing so I could show only `helm` sources, or only sources from `github.com` or only ...

Reply:
The source numbers are not necessary.  They are not in the model too. I was thinking the source list is displayed in the order that they appear in the application manifest. I'll leave it out.

Reply:
> A few things:
> 
> 1. the section (repo url / path / plugin source) should remain on screen while you're scrolling through its content
> 2. headings within it (`Directory`, `Parameters`) should be collapsible
> 3. ideally they should remain visible until their section scrolls off. See the GitHub Pull Request Reviews add-on for Visual Studio Code for an amazing implementation:
>    https://github.com/argoproj/argo-cd/assets/2119212/df478b54-d302-4d5f-868d-ca1478d2b8d2

Regarding points 1 and 3 from above, yeah, the 'distiniguishing' fields should be visible so users won't get lost.  See video.

I'd like to get the READ-ONLY changes in as a first stage. (The URL and PATH fields will move to this new Sources tab from the SUMMARY tab).

The changes to support Editing the parameters will be based on this.

Note that at least the URL and Path fields will possibly be in their own EDITable section within the collapsible Source section.  The reason is that if their values change, then the Type and input parameters will change too.

Reply:
![StickyTitleForSources](https://github.com/argoproj/argo-cd/assets/7726022/27a6fd19-e6a3-42b9-a760-d01903478dc7)

Here I'm taking advantage of the existing white box's additional top space that is already defined in containers.scss.

Reply:
That's not bad -- although it's kinda hard to read. Could you give it a ring and opacity or something so it's easier to see?

Reply:
This is what I was referring to, where if you change the repo URL, or simply change the path within the repo, the input parameters and even the source type can change.  So it's probably better to keep these values as separate sections. So perhaps we will need two Edit buttons, like in the screenshot below? As mentioned, for single source apps, these fields are shown in the Summary section, under a different editable section.  If you change the values there, the parameters tab will be updated once you visit the tab.

<img width="980" alt="multipleEditableSections" src="https://github.com/argoproj/argo-cd/assets/7726022/66d2ca2d-374c-4d80-b481-2e4aae6ba348">

Reply:
Hmmm... Yeah, I think it makes sense to have two edit buttons because they're really *very* unrelated things

Reply:
The initial PR changes for 'read-only' mode is here.  https://github.com/argoproj/argo-cd/pull/17275.  

A lot of it is refactoring. I think we can make more progress from these set of changes.